# Catboost

CatBoost stands for Categorical Boosting. It takes a weak base classifier (a small decision tree) and boosts the weak classifier by learning it many times, each time trying to correct the previous classifier's errors. They are then combined to a boosted classifier. Because the new trees are trained on the difference between true labels and the previous prediction, this method is called gradient boosting. The learning rate will determine the speed of boosting: A small learning rate means that one needs many trees to approximate the training data, but will usually prevent overfitting.

What makes CatBoost a particularly interesting classifier is its handling of categorical values: There is no prior transformation of categorical values necessary (e.g. one-hot-encoding), but CatBoost takes care of all of this by itself. Moreover, CatBoost can combine 2 categorical features to a new one, if this helps the learning process (if e.g. there are strong correlations between two features). It can even do this with continuous features by splitting them into intervals and treating the intervals as categorical information.

Tree-based methods can give a straightforward interpretation of the variable importance because the earlier a variable appears in the split, the more important it is. CatBoost is no exception to this and therefore has its own internal variable importance as an addition to "external" importance estimators like e.g. SHAP.

 <a href="learningsys.org/nips17/assets/papers/paper_11.pdf">Publication on CatBoost</a>



## 1 Installing Catboost and Preparing the Data
We will test CatBoost on the famous Titanic data set. The target variable of the Titanic data is the "Survived" column and the other columns include categorical (e.g. passenger class Pclass and Sex) and numeric (e.g. Age) features. Some of the categorical features are strings, while others are integers. But CatBoost does not bother with that and transforms the categorical features on its own, meaning that we do not have to prepare the categorical features. The test_df has no true labels (y_test) because the data is from a kaggle competition, i.e. you have to estimate the labels yourself.

In [1]:
import sys
!{sys.executable} -m pip install catboost
!{sys.executable} -m pip install scikit-learn
!{sys.executable} -m pip install ipywidgets
!jupyter nbextension enable --py widgetsnbextension

Enabling notebook extension jupyter-js-widgets/extension...
      - Validating: OK


In [2]:
import catboost
from catboost.datasets import titanic
from sklearn.model_selection import train_test_split
from catboost import CatBoostClassifier, Pool, metrics, cv
from sklearn.metrics import accuracy_score
import numpy as np

train_df, test_df = titanic()
print(type(train_df), train_df.shape, test_df.shape)
train_df.head()

<class 'pandas.core.frame.DataFrame'> (891, 12) (418, 11)


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
# Search for missing values
null_value_stats = train_df.isnull().sum(axis=0)
null_value_stats[null_value_stats != 0]

Age         177
Cabin       687
Embarked      2
dtype: int64

As recommended in the official CatBoost tutorial, missing values should be filled with a value way outside of their distribution. Hence, the algorithm will recognise that this is markedly different from the other values.

In [4]:
train_df.fillna(-999, inplace=True)
test_df.fillna(-999, inplace=True)

In [5]:
## Train Test Split

X = train_df.drop('Survived', axis=1)
y = train_df.Survived

X_train, X_validation, y_train, y_validation = train_test_split(X, y, train_size=0.75, random_state=42)
X_test = test_df

# One has to tell CatBoost which features are categorical because it could also treat continuous ones like categorical if one wishes to do so
categorical_features_indices = np.where(X.dtypes != float)[0]

In [6]:
print(y_train.mean(), y_validation.mean())

0.3787425149700599 0.3991031390134529


Because the data is split fairly even in a 40:60 ratio, simply "predicting" that everyone survived or everyone drowned will not result in a good classifier. Hence, using accuracy as a metric can be justified.

## 2 Training the Model

In [7]:
model = CatBoostClassifier(
    custom_loss=[metrics.Accuracy()],
    random_seed=42,
    logging_level='Silent'
)

In [8]:
model.fit(
    X_train, y_train,
    cat_features=categorical_features_indices,
    eval_set=(X_validation, y_validation),
    logging_level='Verbose',  # you can uncomment this for text output, this prints status of training process
    plot=True
)

MetricVisualizer(layout=Layout(align_self='stretch', height='500px'))

Learning rate set to 0.028683
0:	learn: 0.6739988	test: 0.6742630	best: 0.6742630 (0)	total: 62.9ms	remaining: 1m 2s
1:	learn: 0.6589013	test: 0.6592240	best: 0.6592240 (1)	total: 65.6ms	remaining: 32.7s
2:	learn: 0.6421502	test: 0.6426778	best: 0.6426778 (2)	total: 69.9ms	remaining: 23.2s
3:	learn: 0.6297276	test: 0.6302310	best: 0.6302310 (3)	total: 73.1ms	remaining: 18.2s
4:	learn: 0.6147184	test: 0.6198228	best: 0.6198228 (4)	total: 77.3ms	remaining: 15.4s
5:	learn: 0.6017730	test: 0.6073627	best: 0.6073627 (5)	total: 79.4ms	remaining: 13.2s
6:	learn: 0.5885309	test: 0.5956000	best: 0.5956000 (6)	total: 81.4ms	remaining: 11.6s
7:	learn: 0.5783200	test: 0.5858523	best: 0.5858523 (7)	total: 83.7ms	remaining: 10.4s
8:	learn: 0.5665895	test: 0.5743842	best: 0.5743842 (8)	total: 87.3ms	remaining: 9.61s
9:	learn: 0.5575381	test: 0.5662283	best: 0.5662283 (9)	total: 89.6ms	remaining: 8.87s
10:	learn: 0.5491045	test: 0.5575176	best: 0.5575176 (10)	total: 92.2ms	remaining: 8.29s
11:	learn: 

111:	learn: 0.3646549	test: 0.4087192	best: 0.4087192 (111)	total: 350ms	remaining: 2.77s
112:	learn: 0.3634674	test: 0.4075590	best: 0.4075590 (112)	total: 352ms	remaining: 2.76s
113:	learn: 0.3630065	test: 0.4073434	best: 0.4073434 (113)	total: 355ms	remaining: 2.76s
114:	learn: 0.3624458	test: 0.4074634	best: 0.4073434 (113)	total: 357ms	remaining: 2.75s
115:	learn: 0.3616972	test: 0.4071796	best: 0.4071796 (115)	total: 360ms	remaining: 2.74s
116:	learn: 0.3615271	test: 0.4071387	best: 0.4071387 (116)	total: 362ms	remaining: 2.73s
117:	learn: 0.3614277	test: 0.4072280	best: 0.4071387 (116)	total: 364ms	remaining: 2.72s
118:	learn: 0.3603144	test: 0.4062959	best: 0.4062959 (118)	total: 366ms	remaining: 2.71s
119:	learn: 0.3599630	test: 0.4062330	best: 0.4062330 (119)	total: 368ms	remaining: 2.7s
120:	learn: 0.3592807	test: 0.4058909	best: 0.4058909 (120)	total: 371ms	remaining: 2.69s
121:	learn: 0.3587751	test: 0.4056309	best: 0.4056309 (121)	total: 375ms	remaining: 2.7s
122:	learn: 

245:	learn: 0.3120558	test: 0.3947056	best: 0.3945455 (239)	total: 677ms	remaining: 2.08s
246:	learn: 0.3118030	test: 0.3945753	best: 0.3945455 (239)	total: 680ms	remaining: 2.07s
247:	learn: 0.3115541	test: 0.3945034	best: 0.3945034 (247)	total: 682ms	remaining: 2.07s
248:	learn: 0.3113836	test: 0.3945553	best: 0.3945034 (247)	total: 684ms	remaining: 2.06s
249:	learn: 0.3109114	test: 0.3945776	best: 0.3945034 (247)	total: 687ms	remaining: 2.06s
250:	learn: 0.3105904	test: 0.3946490	best: 0.3945034 (247)	total: 689ms	remaining: 2.06s
251:	learn: 0.3104614	test: 0.3946911	best: 0.3945034 (247)	total: 692ms	remaining: 2.05s
252:	learn: 0.3102894	test: 0.3946865	best: 0.3945034 (247)	total: 694ms	remaining: 2.05s
253:	learn: 0.3098849	test: 0.3944783	best: 0.3944783 (253)	total: 696ms	remaining: 2.04s
254:	learn: 0.3096090	test: 0.3945152	best: 0.3944783 (253)	total: 698ms	remaining: 2.04s
255:	learn: 0.3095389	test: 0.3945467	best: 0.3944783 (253)	total: 701ms	remaining: 2.04s
256:	learn

344:	learn: 0.2816979	test: 0.3927250	best: 0.3925689 (343)	total: 930ms	remaining: 1.77s
345:	learn: 0.2812803	test: 0.3932255	best: 0.3925689 (343)	total: 934ms	remaining: 1.76s
346:	learn: 0.2803450	test: 0.3933986	best: 0.3925689 (343)	total: 937ms	remaining: 1.76s
347:	learn: 0.2796085	test: 0.3934529	best: 0.3925689 (343)	total: 947ms	remaining: 1.77s
348:	learn: 0.2793429	test: 0.3935490	best: 0.3925689 (343)	total: 950ms	remaining: 1.77s
349:	learn: 0.2786006	test: 0.3937183	best: 0.3925689 (343)	total: 960ms	remaining: 1.78s
350:	learn: 0.2777571	test: 0.3940834	best: 0.3925689 (343)	total: 963ms	remaining: 1.78s
351:	learn: 0.2776090	test: 0.3940769	best: 0.3925689 (343)	total: 968ms	remaining: 1.78s
352:	learn: 0.2767992	test: 0.3934999	best: 0.3925689 (343)	total: 974ms	remaining: 1.78s
353:	learn: 0.2764869	test: 0.3933342	best: 0.3925689 (343)	total: 977ms	remaining: 1.78s
354:	learn: 0.2757839	test: 0.3935131	best: 0.3925689 (343)	total: 980ms	remaining: 1.78s
355:	learn

455:	learn: 0.2430270	test: 0.3983865	best: 0.3925689 (343)	total: 1.26s	remaining: 1.51s
456:	learn: 0.2427739	test: 0.3985092	best: 0.3925689 (343)	total: 1.27s	remaining: 1.51s
457:	learn: 0.2425688	test: 0.3986027	best: 0.3925689 (343)	total: 1.27s	remaining: 1.5s
458:	learn: 0.2424027	test: 0.3987649	best: 0.3925689 (343)	total: 1.27s	remaining: 1.5s
459:	learn: 0.2420583	test: 0.3987830	best: 0.3925689 (343)	total: 1.28s	remaining: 1.5s
460:	learn: 0.2415432	test: 0.3981029	best: 0.3925689 (343)	total: 1.28s	remaining: 1.5s
461:	learn: 0.2412623	test: 0.3977599	best: 0.3925689 (343)	total: 1.28s	remaining: 1.49s
462:	learn: 0.2409589	test: 0.3981317	best: 0.3925689 (343)	total: 1.29s	remaining: 1.49s
463:	learn: 0.2406763	test: 0.3979839	best: 0.3925689 (343)	total: 1.29s	remaining: 1.49s
464:	learn: 0.2401860	test: 0.3983407	best: 0.3925689 (343)	total: 1.29s	remaining: 1.49s
465:	learn: 0.2398564	test: 0.3984807	best: 0.3925689 (343)	total: 1.29s	remaining: 1.48s
466:	learn: 0.

556:	learn: 0.2156300	test: 0.4035766	best: 0.3925689 (343)	total: 1.66s	remaining: 1.32s
557:	learn: 0.2154524	test: 0.4032750	best: 0.3925689 (343)	total: 1.66s	remaining: 1.32s
558:	learn: 0.2153958	test: 0.4033683	best: 0.3925689 (343)	total: 1.67s	remaining: 1.31s
559:	learn: 0.2153603	test: 0.4033591	best: 0.3925689 (343)	total: 1.67s	remaining: 1.31s
560:	learn: 0.2153034	test: 0.4033728	best: 0.3925689 (343)	total: 1.67s	remaining: 1.31s
561:	learn: 0.2152633	test: 0.4032866	best: 0.3925689 (343)	total: 1.67s	remaining: 1.3s
562:	learn: 0.2148556	test: 0.4036073	best: 0.3925689 (343)	total: 1.68s	remaining: 1.3s
563:	learn: 0.2148436	test: 0.4035961	best: 0.3925689 (343)	total: 1.68s	remaining: 1.3s
564:	learn: 0.2148228	test: 0.4036479	best: 0.3925689 (343)	total: 1.68s	remaining: 1.3s
565:	learn: 0.2147965	test: 0.4035684	best: 0.3925689 (343)	total: 1.69s	remaining: 1.29s
566:	learn: 0.2146438	test: 0.4035601	best: 0.3925689 (343)	total: 1.69s	remaining: 1.29s
567:	learn: 0.

657:	learn: 0.1962348	test: 0.4091966	best: 0.3925689 (343)	total: 1.92s	remaining: 998ms
658:	learn: 0.1961597	test: 0.4091615	best: 0.3925689 (343)	total: 1.92s	remaining: 995ms
659:	learn: 0.1956394	test: 0.4092077	best: 0.3925689 (343)	total: 1.93s	remaining: 992ms
660:	learn: 0.1956237	test: 0.4092356	best: 0.3925689 (343)	total: 1.94s	remaining: 996ms
661:	learn: 0.1954545	test: 0.4093616	best: 0.3925689 (343)	total: 1.96s	remaining: 999ms
662:	learn: 0.1951190	test: 0.4097803	best: 0.3925689 (343)	total: 1.96s	remaining: 996ms
663:	learn: 0.1950099	test: 0.4097903	best: 0.3925689 (343)	total: 1.97s	remaining: 996ms
664:	learn: 0.1948989	test: 0.4097491	best: 0.3925689 (343)	total: 1.97s	remaining: 992ms
665:	learn: 0.1945474	test: 0.4100763	best: 0.3925689 (343)	total: 1.97s	remaining: 989ms
666:	learn: 0.1944861	test: 0.4100757	best: 0.3925689 (343)	total: 1.98s	remaining: 986ms
667:	learn: 0.1943174	test: 0.4101521	best: 0.3925689 (343)	total: 1.98s	remaining: 983ms
668:	learn

765:	learn: 0.1753264	test: 0.4165993	best: 0.3925689 (343)	total: 2.26s	remaining: 690ms
766:	learn: 0.1752084	test: 0.4166235	best: 0.3925689 (343)	total: 2.26s	remaining: 687ms
767:	learn: 0.1749531	test: 0.4170032	best: 0.3925689 (343)	total: 2.26s	remaining: 684ms
768:	learn: 0.1749129	test: 0.4170263	best: 0.3925689 (343)	total: 2.27s	remaining: 681ms
769:	learn: 0.1745286	test: 0.4176999	best: 0.3925689 (343)	total: 2.27s	remaining: 678ms
770:	learn: 0.1745228	test: 0.4177098	best: 0.3925689 (343)	total: 2.27s	remaining: 675ms
771:	learn: 0.1741362	test: 0.4176875	best: 0.3925689 (343)	total: 2.28s	remaining: 673ms
772:	learn: 0.1737494	test: 0.4178495	best: 0.3925689 (343)	total: 2.28s	remaining: 670ms
773:	learn: 0.1737326	test: 0.4179174	best: 0.3925689 (343)	total: 2.28s	remaining: 667ms
774:	learn: 0.1735426	test: 0.4178826	best: 0.3925689 (343)	total: 2.29s	remaining: 663ms
775:	learn: 0.1732704	test: 0.4181271	best: 0.3925689 (343)	total: 2.29s	remaining: 660ms
776:	learn

868:	learn: 0.1584100	test: 0.4254165	best: 0.3925689 (343)	total: 2.54s	remaining: 383ms
869:	learn: 0.1582611	test: 0.4252118	best: 0.3925689 (343)	total: 2.54s	remaining: 380ms
870:	learn: 0.1580364	test: 0.4255225	best: 0.3925689 (343)	total: 2.54s	remaining: 377ms
871:	learn: 0.1577863	test: 0.4264693	best: 0.3925689 (343)	total: 2.55s	remaining: 374ms
872:	learn: 0.1575010	test: 0.4266996	best: 0.3925689 (343)	total: 2.55s	remaining: 371ms
873:	learn: 0.1572197	test: 0.4268858	best: 0.3925689 (343)	total: 2.55s	remaining: 368ms
874:	learn: 0.1571717	test: 0.4268197	best: 0.3925689 (343)	total: 2.55s	remaining: 365ms
875:	learn: 0.1570976	test: 0.4269053	best: 0.3925689 (343)	total: 2.56s	remaining: 362ms
876:	learn: 0.1569863	test: 0.4270429	best: 0.3925689 (343)	total: 2.56s	remaining: 359ms
877:	learn: 0.1566991	test: 0.4269143	best: 0.3925689 (343)	total: 2.56s	remaining: 356ms
878:	learn: 0.1566554	test: 0.4270517	best: 0.3925689 (343)	total: 2.56s	remaining: 353ms
879:	learn

980:	learn: 0.1404468	test: 0.4366938	best: 0.3925689 (343)	total: 2.84s	remaining: 55.1ms
981:	learn: 0.1403263	test: 0.4366148	best: 0.3925689 (343)	total: 2.85s	remaining: 52.2ms
982:	learn: 0.1402955	test: 0.4365037	best: 0.3925689 (343)	total: 2.85s	remaining: 49.3ms
983:	learn: 0.1401256	test: 0.4366547	best: 0.3925689 (343)	total: 2.85s	remaining: 46.4ms
984:	learn: 0.1400961	test: 0.4366136	best: 0.3925689 (343)	total: 2.86s	remaining: 43.5ms
985:	learn: 0.1398101	test: 0.4368906	best: 0.3925689 (343)	total: 2.86s	remaining: 40.6ms
986:	learn: 0.1396404	test: 0.4370100	best: 0.3925689 (343)	total: 2.86s	remaining: 37.7ms
987:	learn: 0.1393904	test: 0.4370537	best: 0.3925689 (343)	total: 2.87s	remaining: 34.8ms
988:	learn: 0.1392551	test: 0.4371019	best: 0.3925689 (343)	total: 2.87s	remaining: 31.9ms
989:	learn: 0.1390832	test: 0.4368715	best: 0.3925689 (343)	total: 2.87s	remaining: 29ms
990:	learn: 0.1390399	test: 0.4367513	best: 0.3925689 (343)	total: 2.87s	remaining: 26.1ms
9

One can see that after 344 iterations, further boosting leads to a worse generalisation error, i.e. to overfitting. Hence, CatBoost chooses the 344th iteration as the best model. We use that model as a starting point for cross validation. Note that CatBoosts cross validation wants a catboost.core.Pool object as input that combines the features X, the target y and the categorical index.

In [9]:
cv_params = model.get_params()
cv_params.update({
    'loss_function': metrics.Logloss()
})
cv_data = cv(
    Pool(X, y, cat_features=categorical_features_indices),
    cv_params,
    fold_count = 4, # number of cross validations
    logging_level="Verbose" # tells us about the intermediary results
)

Training on fold [0/4]
0:	learn: 0.6761669	test: 0.6768083	best: 0.6768083 (0)	total: 18.6ms	remaining: 18.6s
1:	learn: 0.6583334	test: 0.6608511	best: 0.6608511 (1)	total: 27.1ms	remaining: 13.5s
2:	learn: 0.6423837	test: 0.6461866	best: 0.6461866 (2)	total: 34.5ms	remaining: 11.5s
3:	learn: 0.6283679	test: 0.6329292	best: 0.6329292 (3)	total: 39ms	remaining: 9.72s
4:	learn: 0.6144513	test: 0.6195135	best: 0.6195135 (4)	total: 43.6ms	remaining: 8.67s
5:	learn: 0.6026594	test: 0.6083604	best: 0.6083604 (5)	total: 48.9ms	remaining: 8.1s
6:	learn: 0.5919314	test: 0.5974993	best: 0.5974993 (6)	total: 51.6ms	remaining: 7.32s
7:	learn: 0.5802086	test: 0.5875367	best: 0.5875367 (7)	total: 56.7ms	remaining: 7.03s
8:	learn: 0.5671302	test: 0.5774444	best: 0.5774444 (8)	total: 61.5ms	remaining: 6.77s
9:	learn: 0.5613196	test: 0.5718529	best: 0.5718529 (9)	total: 63ms	remaining: 6.24s
10:	learn: 0.5524883	test: 0.5634700	best: 0.5634700 (10)	total: 70.9ms	remaining: 6.37s
11:	learn: 0.5420783	te

97:	learn: 0.3561392	test: 0.4216589	best: 0.4214465 (95)	total: 524ms	remaining: 4.82s
98:	learn: 0.3552380	test: 0.4212137	best: 0.4212137 (98)	total: 529ms	remaining: 4.82s
99:	learn: 0.3547494	test: 0.4210964	best: 0.4210964 (99)	total: 533ms	remaining: 4.8s
100:	learn: 0.3544346	test: 0.4208324	best: 0.4208324 (100)	total: 536ms	remaining: 4.77s
101:	learn: 0.3537253	test: 0.4208944	best: 0.4208324 (100)	total: 543ms	remaining: 4.78s
102:	learn: 0.3529604	test: 0.4208832	best: 0.4208324 (100)	total: 548ms	remaining: 4.77s
103:	learn: 0.3519260	test: 0.4206761	best: 0.4206761 (103)	total: 557ms	remaining: 4.8s
104:	learn: 0.3513192	test: 0.4205804	best: 0.4205804 (104)	total: 562ms	remaining: 4.79s
105:	learn: 0.3506273	test: 0.4207990	best: 0.4205804 (104)	total: 571ms	remaining: 4.81s
106:	learn: 0.3498566	test: 0.4208351	best: 0.4205804 (104)	total: 575ms	remaining: 4.8s
107:	learn: 0.3492378	test: 0.4206773	best: 0.4205804 (104)	total: 579ms	remaining: 4.78s
108:	learn: 0.34839

198:	learn: 0.3099827	test: 0.4155047	best: 0.4151780 (185)	total: 1.07s	remaining: 4.31s
199:	learn: 0.3099378	test: 0.4155051	best: 0.4151780 (185)	total: 1.07s	remaining: 4.3s
200:	learn: 0.3096826	test: 0.4156289	best: 0.4151780 (185)	total: 1.08s	remaining: 4.31s
201:	learn: 0.3094587	test: 0.4155428	best: 0.4151780 (185)	total: 1.09s	remaining: 4.32s
202:	learn: 0.3091965	test: 0.4155100	best: 0.4151780 (185)	total: 1.1s	remaining: 4.31s
203:	learn: 0.3089981	test: 0.4151900	best: 0.4151780 (185)	total: 1.1s	remaining: 4.31s
204:	learn: 0.3085965	test: 0.4150236	best: 0.4150236 (204)	total: 1.11s	remaining: 4.29s
205:	learn: 0.3082000	test: 0.4155330	best: 0.4150236 (204)	total: 1.11s	remaining: 4.28s
206:	learn: 0.3081189	test: 0.4154364	best: 0.4150236 (204)	total: 1.12s	remaining: 4.29s
207:	learn: 0.3076681	test: 0.4154269	best: 0.4150236 (204)	total: 1.12s	remaining: 4.28s
208:	learn: 0.3076370	test: 0.4154412	best: 0.4150236 (204)	total: 1.13s	remaining: 4.26s
209:	learn: 0

294:	learn: 0.2773919	test: 0.4167648	best: 0.4150236 (204)	total: 1.63s	remaining: 3.9s
295:	learn: 0.2773428	test: 0.4167625	best: 0.4150236 (204)	total: 1.64s	remaining: 3.9s
296:	learn: 0.2770425	test: 0.4166856	best: 0.4150236 (204)	total: 1.64s	remaining: 3.89s
297:	learn: 0.2766801	test: 0.4167605	best: 0.4150236 (204)	total: 1.65s	remaining: 3.88s
298:	learn: 0.2763049	test: 0.4169842	best: 0.4150236 (204)	total: 1.65s	remaining: 3.88s
299:	learn: 0.2759852	test: 0.4173463	best: 0.4150236 (204)	total: 1.66s	remaining: 3.87s
300:	learn: 0.2756637	test: 0.4174560	best: 0.4150236 (204)	total: 1.66s	remaining: 3.86s
301:	learn: 0.2751539	test: 0.4174426	best: 0.4150236 (204)	total: 1.67s	remaining: 3.85s
302:	learn: 0.2748412	test: 0.4173882	best: 0.4150236 (204)	total: 1.67s	remaining: 3.85s
303:	learn: 0.2746520	test: 0.4174950	best: 0.4150236 (204)	total: 1.68s	remaining: 3.84s
304:	learn: 0.2741592	test: 0.4173315	best: 0.4150236 (204)	total: 1.68s	remaining: 3.83s
305:	learn: 

389:	learn: 0.2465044	test: 0.4192620	best: 0.4150236 (204)	total: 2.2s	remaining: 3.44s
390:	learn: 0.2461573	test: 0.4194118	best: 0.4150236 (204)	total: 2.21s	remaining: 3.44s
391:	learn: 0.2458863	test: 0.4196550	best: 0.4150236 (204)	total: 2.21s	remaining: 3.43s
392:	learn: 0.2458480	test: 0.4197132	best: 0.4150236 (204)	total: 2.22s	remaining: 3.42s
393:	learn: 0.2455651	test: 0.4197800	best: 0.4150236 (204)	total: 2.22s	remaining: 3.42s
394:	learn: 0.2449520	test: 0.4196968	best: 0.4150236 (204)	total: 2.23s	remaining: 3.41s
395:	learn: 0.2442865	test: 0.4193730	best: 0.4150236 (204)	total: 2.23s	remaining: 3.4s
396:	learn: 0.2441380	test: 0.4195962	best: 0.4150236 (204)	total: 2.24s	remaining: 3.4s
397:	learn: 0.2439957	test: 0.4198873	best: 0.4150236 (204)	total: 2.24s	remaining: 3.39s
398:	learn: 0.2438195	test: 0.4198992	best: 0.4150236 (204)	total: 2.25s	remaining: 3.38s
399:	learn: 0.2434223	test: 0.4200363	best: 0.4150236 (204)	total: 2.26s	remaining: 3.38s
400:	learn: 0

490:	learn: 0.2179166	test: 0.4245206	best: 0.4150236 (204)	total: 2.83s	remaining: 2.94s
491:	learn: 0.2178768	test: 0.4246152	best: 0.4150236 (204)	total: 2.84s	remaining: 2.93s
492:	learn: 0.2176153	test: 0.4247964	best: 0.4150236 (204)	total: 2.85s	remaining: 2.93s
493:	learn: 0.2173890	test: 0.4245737	best: 0.4150236 (204)	total: 2.85s	remaining: 2.92s
494:	learn: 0.2168391	test: 0.4244805	best: 0.4150236 (204)	total: 2.86s	remaining: 2.92s
495:	learn: 0.2166951	test: 0.4246764	best: 0.4150236 (204)	total: 2.87s	remaining: 2.92s
496:	learn: 0.2166123	test: 0.4247455	best: 0.4150236 (204)	total: 2.87s	remaining: 2.91s
497:	learn: 0.2165169	test: 0.4248671	best: 0.4150236 (204)	total: 2.88s	remaining: 2.9s
498:	learn: 0.2159213	test: 0.4247371	best: 0.4150236 (204)	total: 2.88s	remaining: 2.9s
499:	learn: 0.2156345	test: 0.4248715	best: 0.4150236 (204)	total: 2.89s	remaining: 2.89s
500:	learn: 0.2153456	test: 0.4254340	best: 0.4150236 (204)	total: 2.9s	remaining: 2.89s
501:	learn: 0

586:	learn: 0.1939059	test: 0.4273812	best: 0.4150236 (204)	total: 3.51s	remaining: 2.47s
587:	learn: 0.1937605	test: 0.4275104	best: 0.4150236 (204)	total: 3.52s	remaining: 2.47s
588:	learn: 0.1935403	test: 0.4278063	best: 0.4150236 (204)	total: 3.53s	remaining: 2.46s
589:	learn: 0.1931130	test: 0.4275223	best: 0.4150236 (204)	total: 3.54s	remaining: 2.46s
590:	learn: 0.1930429	test: 0.4275115	best: 0.4150236 (204)	total: 3.55s	remaining: 2.46s
591:	learn: 0.1928453	test: 0.4274579	best: 0.4150236 (204)	total: 3.55s	remaining: 2.45s
592:	learn: 0.1924668	test: 0.4273625	best: 0.4150236 (204)	total: 3.56s	remaining: 2.44s
593:	learn: 0.1920739	test: 0.4274170	best: 0.4150236 (204)	total: 3.57s	remaining: 2.44s
594:	learn: 0.1917396	test: 0.4273336	best: 0.4150236 (204)	total: 3.58s	remaining: 2.43s
595:	learn: 0.1915415	test: 0.4275588	best: 0.4150236 (204)	total: 3.59s	remaining: 2.43s
596:	learn: 0.1914592	test: 0.4273493	best: 0.4150236 (204)	total: 3.6s	remaining: 2.43s
597:	learn:

684:	learn: 0.1743767	test: 0.4318440	best: 0.4150236 (204)	total: 4.12s	remaining: 1.89s
685:	learn: 0.1741651	test: 0.4320257	best: 0.4150236 (204)	total: 4.13s	remaining: 1.89s
686:	learn: 0.1741008	test: 0.4320780	best: 0.4150236 (204)	total: 4.14s	remaining: 1.89s
687:	learn: 0.1740200	test: 0.4317772	best: 0.4150236 (204)	total: 4.14s	remaining: 1.88s
688:	learn: 0.1737514	test: 0.4320844	best: 0.4150236 (204)	total: 4.15s	remaining: 1.87s
689:	learn: 0.1735033	test: 0.4322464	best: 0.4150236 (204)	total: 4.15s	remaining: 1.87s
690:	learn: 0.1733387	test: 0.4322793	best: 0.4150236 (204)	total: 4.16s	remaining: 1.86s
691:	learn: 0.1730234	test: 0.4324741	best: 0.4150236 (204)	total: 4.17s	remaining: 1.85s
692:	learn: 0.1728925	test: 0.4325393	best: 0.4150236 (204)	total: 4.18s	remaining: 1.85s
693:	learn: 0.1727615	test: 0.4327565	best: 0.4150236 (204)	total: 4.18s	remaining: 1.84s
694:	learn: 0.1725782	test: 0.4327177	best: 0.4150236 (204)	total: 4.19s	remaining: 1.84s
695:	learn

778:	learn: 0.1604598	test: 0.4370277	best: 0.4150236 (204)	total: 4.66s	remaining: 1.32s
779:	learn: 0.1603165	test: 0.4369764	best: 0.4150236 (204)	total: 4.67s	remaining: 1.32s
780:	learn: 0.1600688	test: 0.4375237	best: 0.4150236 (204)	total: 4.67s	remaining: 1.31s
781:	learn: 0.1600407	test: 0.4376014	best: 0.4150236 (204)	total: 4.69s	remaining: 1.31s
782:	learn: 0.1599000	test: 0.4378392	best: 0.4150236 (204)	total: 4.69s	remaining: 1.3s
783:	learn: 0.1598555	test: 0.4378845	best: 0.4150236 (204)	total: 4.69s	remaining: 1.29s
784:	learn: 0.1595530	test: 0.4379366	best: 0.4150236 (204)	total: 4.7s	remaining: 1.29s
785:	learn: 0.1594137	test: 0.4378505	best: 0.4150236 (204)	total: 4.7s	remaining: 1.28s
786:	learn: 0.1592928	test: 0.4380875	best: 0.4150236 (204)	total: 4.71s	remaining: 1.27s
787:	learn: 0.1591794	test: 0.4382460	best: 0.4150236 (204)	total: 4.72s	remaining: 1.27s
788:	learn: 0.1590761	test: 0.4383818	best: 0.4150236 (204)	total: 4.72s	remaining: 1.26s
789:	learn: 0

896:	learn: 0.1429598	test: 0.4404560	best: 0.4150236 (204)	total: 5.37s	remaining: 617ms
897:	learn: 0.1427459	test: 0.4406437	best: 0.4150236 (204)	total: 5.38s	remaining: 611ms
898:	learn: 0.1426594	test: 0.4405525	best: 0.4150236 (204)	total: 5.39s	remaining: 605ms
899:	learn: 0.1424228	test: 0.4407016	best: 0.4150236 (204)	total: 5.39s	remaining: 599ms
900:	learn: 0.1423710	test: 0.4410657	best: 0.4150236 (204)	total: 5.4s	remaining: 593ms
901:	learn: 0.1421183	test: 0.4410545	best: 0.4150236 (204)	total: 5.41s	remaining: 587ms
902:	learn: 0.1417865	test: 0.4408406	best: 0.4150236 (204)	total: 5.42s	remaining: 582ms
903:	learn: 0.1416688	test: 0.4409089	best: 0.4150236 (204)	total: 5.42s	remaining: 576ms
904:	learn: 0.1415968	test: 0.4407734	best: 0.4150236 (204)	total: 5.44s	remaining: 571ms
905:	learn: 0.1415213	test: 0.4409002	best: 0.4150236 (204)	total: 5.44s	remaining: 565ms
906:	learn: 0.1413828	test: 0.4408461	best: 0.4150236 (204)	total: 5.45s	remaining: 559ms
907:	learn:

988:	learn: 0.1301010	test: 0.4446110	best: 0.4150236 (204)	total: 5.92s	remaining: 65.8ms
989:	learn: 0.1300026	test: 0.4449098	best: 0.4150236 (204)	total: 5.92s	remaining: 59.9ms
990:	learn: 0.1298669	test: 0.4449175	best: 0.4150236 (204)	total: 5.94s	remaining: 53.9ms
991:	learn: 0.1297614	test: 0.4451571	best: 0.4150236 (204)	total: 5.95s	remaining: 48ms
992:	learn: 0.1296004	test: 0.4454056	best: 0.4150236 (204)	total: 5.95s	remaining: 42ms
993:	learn: 0.1295287	test: 0.4456059	best: 0.4150236 (204)	total: 5.96s	remaining: 36ms
994:	learn: 0.1294776	test: 0.4454814	best: 0.4150236 (204)	total: 5.97s	remaining: 30ms
995:	learn: 0.1293024	test: 0.4454830	best: 0.4150236 (204)	total: 5.97s	remaining: 24ms
996:	learn: 0.1290621	test: 0.4456517	best: 0.4150236 (204)	total: 5.98s	remaining: 18ms
997:	learn: 0.1289485	test: 0.4455689	best: 0.4150236 (204)	total: 5.99s	remaining: 12ms
998:	learn: 0.1288068	test: 0.4453811	best: 0.4150236 (204)	total: 6s	remaining: 6ms
999:	learn: 0.12872

86:	learn: 0.3810605	test: 0.4019480	best: 0.4017808 (85)	total: 689ms	remaining: 7.23s
87:	learn: 0.3804860	test: 0.4016431	best: 0.4016431 (87)	total: 698ms	remaining: 7.23s
88:	learn: 0.3803059	test: 0.4015724	best: 0.4015724 (88)	total: 704ms	remaining: 7.21s
89:	learn: 0.3790209	test: 0.4005326	best: 0.4005326 (89)	total: 718ms	remaining: 7.26s
90:	learn: 0.3782971	test: 0.4001379	best: 0.4001379 (90)	total: 728ms	remaining: 7.27s
91:	learn: 0.3776928	test: 0.3994293	best: 0.3994293 (91)	total: 734ms	remaining: 7.24s
92:	learn: 0.3768989	test: 0.3993915	best: 0.3993915 (92)	total: 742ms	remaining: 7.23s
93:	learn: 0.3768958	test: 0.3993758	best: 0.3993758 (93)	total: 744ms	remaining: 7.17s
94:	learn: 0.3765474	test: 0.3975670	best: 0.3975670 (94)	total: 768ms	remaining: 7.31s
95:	learn: 0.3759006	test: 0.3969885	best: 0.3969885 (95)	total: 787ms	remaining: 7.41s
96:	learn: 0.3752560	test: 0.3962467	best: 0.3962467 (96)	total: 805ms	remaining: 7.49s
97:	learn: 0.3752050	test: 0.396

178:	learn: 0.3410969	test: 0.3830032	best: 0.3830032 (178)	total: 1.51s	remaining: 6.94s
179:	learn: 0.3408303	test: 0.3827815	best: 0.3827815 (179)	total: 1.52s	remaining: 6.93s
180:	learn: 0.3399877	test: 0.3823588	best: 0.3823588 (180)	total: 1.55s	remaining: 7.03s
181:	learn: 0.3395145	test: 0.3823304	best: 0.3823304 (181)	total: 1.56s	remaining: 7s
182:	learn: 0.3389036	test: 0.3822943	best: 0.3822943 (182)	total: 1.57s	remaining: 7.03s
183:	learn: 0.3384207	test: 0.3820152	best: 0.3820152 (183)	total: 1.58s	remaining: 7.02s
184:	learn: 0.3377125	test: 0.3814684	best: 0.3814684 (184)	total: 1.59s	remaining: 7s
185:	learn: 0.3376010	test: 0.3816400	best: 0.3814684 (184)	total: 1.59s	remaining: 6.98s
186:	learn: 0.3363601	test: 0.3814244	best: 0.3814244 (186)	total: 1.61s	remaining: 6.99s
187:	learn: 0.3360675	test: 0.3811952	best: 0.3811952 (187)	total: 1.62s	remaining: 7.01s
188:	learn: 0.3359979	test: 0.3812846	best: 0.3811952 (187)	total: 1.63s	remaining: 6.98s
189:	learn: 0.33

270:	learn: 0.3129588	test: 0.3765667	best: 0.3765048 (260)	total: 2.14s	remaining: 5.76s
271:	learn: 0.3124669	test: 0.3761958	best: 0.3761958 (271)	total: 2.15s	remaining: 5.75s
272:	learn: 0.3123785	test: 0.3763446	best: 0.3761958 (271)	total: 2.16s	remaining: 5.75s
273:	learn: 0.3118736	test: 0.3763087	best: 0.3761958 (271)	total: 2.16s	remaining: 5.73s
274:	learn: 0.3116273	test: 0.3762392	best: 0.3761958 (271)	total: 2.17s	remaining: 5.71s
275:	learn: 0.3115735	test: 0.3761450	best: 0.3761450 (275)	total: 2.17s	remaining: 5.7s
276:	learn: 0.3114571	test: 0.3760392	best: 0.3760392 (276)	total: 2.18s	remaining: 5.68s
277:	learn: 0.3113802	test: 0.3758971	best: 0.3758971 (277)	total: 2.18s	remaining: 5.67s
278:	learn: 0.3111330	test: 0.3757989	best: 0.3757989 (278)	total: 2.2s	remaining: 5.68s
279:	learn: 0.3110860	test: 0.3756795	best: 0.3756795 (279)	total: 2.2s	remaining: 5.66s
280:	learn: 0.3110121	test: 0.3756245	best: 0.3756245 (280)	total: 2.21s	remaining: 5.65s
281:	learn: 0

379:	learn: 0.2872962	test: 0.3746501	best: 0.3731240 (348)	total: 2.74s	remaining: 4.48s
380:	learn: 0.2871209	test: 0.3744075	best: 0.3731240 (348)	total: 2.75s	remaining: 4.46s
381:	learn: 0.2869492	test: 0.3746455	best: 0.3731240 (348)	total: 2.75s	remaining: 4.45s
382:	learn: 0.2863564	test: 0.3748280	best: 0.3731240 (348)	total: 2.76s	remaining: 4.44s
383:	learn: 0.2861292	test: 0.3748121	best: 0.3731240 (348)	total: 2.76s	remaining: 4.43s
384:	learn: 0.2857478	test: 0.3750064	best: 0.3731240 (348)	total: 2.77s	remaining: 4.42s
385:	learn: 0.2852939	test: 0.3749629	best: 0.3731240 (348)	total: 2.77s	remaining: 4.41s
386:	learn: 0.2847724	test: 0.3750133	best: 0.3731240 (348)	total: 2.78s	remaining: 4.4s
387:	learn: 0.2845977	test: 0.3750686	best: 0.3731240 (348)	total: 2.78s	remaining: 4.39s
388:	learn: 0.2845092	test: 0.3751214	best: 0.3731240 (348)	total: 2.79s	remaining: 4.38s
389:	learn: 0.2843637	test: 0.3750175	best: 0.3731240 (348)	total: 2.79s	remaining: 4.37s
390:	learn:

477:	learn: 0.2590251	test: 0.3780824	best: 0.3731240 (348)	total: 3.35s	remaining: 3.66s
478:	learn: 0.2588465	test: 0.3777820	best: 0.3731240 (348)	total: 3.35s	remaining: 3.65s
479:	learn: 0.2586126	test: 0.3779304	best: 0.3731240 (348)	total: 3.36s	remaining: 3.64s
480:	learn: 0.2583989	test: 0.3782059	best: 0.3731240 (348)	total: 3.38s	remaining: 3.64s
481:	learn: 0.2578040	test: 0.3780788	best: 0.3731240 (348)	total: 3.38s	remaining: 3.63s
482:	learn: 0.2576394	test: 0.3780803	best: 0.3731240 (348)	total: 3.39s	remaining: 3.63s
483:	learn: 0.2573351	test: 0.3781604	best: 0.3731240 (348)	total: 3.4s	remaining: 3.62s
484:	learn: 0.2567406	test: 0.3782719	best: 0.3731240 (348)	total: 3.4s	remaining: 3.61s
485:	learn: 0.2566759	test: 0.3781640	best: 0.3731240 (348)	total: 3.41s	remaining: 3.61s
486:	learn: 0.2562143	test: 0.3779819	best: 0.3731240 (348)	total: 3.42s	remaining: 3.6s
487:	learn: 0.2558432	test: 0.3780024	best: 0.3731240 (348)	total: 3.43s	remaining: 3.6s
488:	learn: 0.

575:	learn: 0.2356099	test: 0.3793130	best: 0.3731240 (348)	total: 3.95s	remaining: 2.9s
576:	learn: 0.2354157	test: 0.3790997	best: 0.3731240 (348)	total: 3.97s	remaining: 2.91s
577:	learn: 0.2352770	test: 0.3789338	best: 0.3731240 (348)	total: 3.97s	remaining: 2.9s
578:	learn: 0.2344580	test: 0.3789693	best: 0.3731240 (348)	total: 3.98s	remaining: 2.89s
579:	learn: 0.2342464	test: 0.3787094	best: 0.3731240 (348)	total: 3.98s	remaining: 2.88s
580:	learn: 0.2340084	test: 0.3786401	best: 0.3731240 (348)	total: 3.99s	remaining: 2.88s
581:	learn: 0.2335833	test: 0.3788932	best: 0.3731240 (348)	total: 3.99s	remaining: 2.87s
582:	learn: 0.2333787	test: 0.3786613	best: 0.3731240 (348)	total: 4s	remaining: 2.86s
583:	learn: 0.2330058	test: 0.3787132	best: 0.3731240 (348)	total: 4.01s	remaining: 2.85s
584:	learn: 0.2326608	test: 0.3785318	best: 0.3731240 (348)	total: 4.02s	remaining: 2.85s
585:	learn: 0.2323382	test: 0.3788842	best: 0.3731240 (348)	total: 4.03s	remaining: 2.84s
586:	learn: 0.2

680:	learn: 0.2153853	test: 0.3809955	best: 0.3731240 (348)	total: 4.56s	remaining: 2.13s
681:	learn: 0.2152282	test: 0.3808638	best: 0.3731240 (348)	total: 4.56s	remaining: 2.13s
682:	learn: 0.2150692	test: 0.3807270	best: 0.3731240 (348)	total: 4.57s	remaining: 2.12s
683:	learn: 0.2147640	test: 0.3805659	best: 0.3731240 (348)	total: 4.58s	remaining: 2.12s
684:	learn: 0.2146352	test: 0.3809219	best: 0.3731240 (348)	total: 4.59s	remaining: 2.11s
685:	learn: 0.2145640	test: 0.3808894	best: 0.3731240 (348)	total: 4.59s	remaining: 2.1s
686:	learn: 0.2140821	test: 0.3807228	best: 0.3731240 (348)	total: 4.6s	remaining: 2.1s
687:	learn: 0.2135909	test: 0.3808464	best: 0.3731240 (348)	total: 4.61s	remaining: 2.09s
688:	learn: 0.2132938	test: 0.3810920	best: 0.3731240 (348)	total: 4.62s	remaining: 2.08s
689:	learn: 0.2131169	test: 0.3813850	best: 0.3731240 (348)	total: 4.63s	remaining: 2.08s
690:	learn: 0.2130006	test: 0.3811920	best: 0.3731240 (348)	total: 4.63s	remaining: 2.07s
691:	learn: 0

774:	learn: 0.1980666	test: 0.3822852	best: 0.3731240 (348)	total: 5.17s	remaining: 1.5s
775:	learn: 0.1979777	test: 0.3822887	best: 0.3731240 (348)	total: 5.18s	remaining: 1.5s
776:	learn: 0.1978927	test: 0.3822460	best: 0.3731240 (348)	total: 5.19s	remaining: 1.49s
777:	learn: 0.1976755	test: 0.3820215	best: 0.3731240 (348)	total: 5.2s	remaining: 1.48s
778:	learn: 0.1975705	test: 0.3820640	best: 0.3731240 (348)	total: 5.2s	remaining: 1.48s
779:	learn: 0.1975626	test: 0.3820851	best: 0.3731240 (348)	total: 5.23s	remaining: 1.47s
780:	learn: 0.1974759	test: 0.3817898	best: 0.3731240 (348)	total: 5.23s	remaining: 1.47s
781:	learn: 0.1973913	test: 0.3817457	best: 0.3731240 (348)	total: 5.24s	remaining: 1.46s
782:	learn: 0.1972615	test: 0.3816627	best: 0.3731240 (348)	total: 5.25s	remaining: 1.45s
783:	learn: 0.1970241	test: 0.3819092	best: 0.3731240 (348)	total: 5.26s	remaining: 1.45s
784:	learn: 0.1969619	test: 0.3818864	best: 0.3731240 (348)	total: 5.27s	remaining: 1.44s
785:	learn: 0.

886:	learn: 0.1840174	test: 0.3811844	best: 0.3731240 (348)	total: 5.98s	remaining: 761ms
887:	learn: 0.1839696	test: 0.3811486	best: 0.3731240 (348)	total: 5.98s	remaining: 754ms
888:	learn: 0.1838973	test: 0.3811376	best: 0.3731240 (348)	total: 5.99s	remaining: 748ms
889:	learn: 0.1837250	test: 0.3808599	best: 0.3731240 (348)	total: 6s	remaining: 741ms
890:	learn: 0.1836379	test: 0.3808504	best: 0.3731240 (348)	total: 6s	remaining: 734ms
891:	learn: 0.1835190	test: 0.3808972	best: 0.3731240 (348)	total: 6.01s	remaining: 728ms
892:	learn: 0.1833696	test: 0.3807251	best: 0.3731240 (348)	total: 6.01s	remaining: 720ms
893:	learn: 0.1833119	test: 0.3807211	best: 0.3731240 (348)	total: 6.02s	remaining: 714ms
894:	learn: 0.1832961	test: 0.3806995	best: 0.3731240 (348)	total: 6.02s	remaining: 706ms
895:	learn: 0.1832312	test: 0.3807527	best: 0.3731240 (348)	total: 6.03s	remaining: 700ms
896:	learn: 0.1828830	test: 0.3806991	best: 0.3731240 (348)	total: 6.04s	remaining: 693ms
897:	learn: 0.18

987:	learn: 0.1715698	test: 0.3817060	best: 0.3731240 (348)	total: 6.58s	remaining: 79.9ms
988:	learn: 0.1715169	test: 0.3815502	best: 0.3731240 (348)	total: 6.58s	remaining: 73.2ms
989:	learn: 0.1714160	test: 0.3815604	best: 0.3731240 (348)	total: 6.59s	remaining: 66.6ms
990:	learn: 0.1712291	test: 0.3816309	best: 0.3731240 (348)	total: 6.59s	remaining: 59.9ms
991:	learn: 0.1709351	test: 0.3815245	best: 0.3731240 (348)	total: 6.6s	remaining: 53.2ms
992:	learn: 0.1709106	test: 0.3815830	best: 0.3731240 (348)	total: 6.6s	remaining: 46.5ms
993:	learn: 0.1708933	test: 0.3815913	best: 0.3731240 (348)	total: 6.61s	remaining: 39.9ms
994:	learn: 0.1708038	test: 0.3819438	best: 0.3731240 (348)	total: 6.62s	remaining: 33.3ms
995:	learn: 0.1706271	test: 0.3819615	best: 0.3731240 (348)	total: 6.62s	remaining: 26.6ms
996:	learn: 0.1705424	test: 0.3817419	best: 0.3731240 (348)	total: 6.63s	remaining: 20ms
997:	learn: 0.1703502	test: 0.3814011	best: 0.3731240 (348)	total: 6.64s	remaining: 13.3ms
998

82:	learn: 0.3677885	test: 0.4295547	best: 0.4295547 (82)	total: 493ms	remaining: 5.44s
83:	learn: 0.3663917	test: 0.4295047	best: 0.4295047 (83)	total: 499ms	remaining: 5.45s
84:	learn: 0.3660674	test: 0.4291078	best: 0.4291078 (84)	total: 507ms	remaining: 5.46s
85:	learn: 0.3659231	test: 0.4289539	best: 0.4289539 (85)	total: 508ms	remaining: 5.4s
86:	learn: 0.3648838	test: 0.4290588	best: 0.4289539 (85)	total: 514ms	remaining: 5.39s
87:	learn: 0.3625027	test: 0.4272956	best: 0.4272956 (87)	total: 519ms	remaining: 5.38s
88:	learn: 0.3617272	test: 0.4271366	best: 0.4271366 (88)	total: 523ms	remaining: 5.35s
89:	learn: 0.3607647	test: 0.4268863	best: 0.4268863 (89)	total: 527ms	remaining: 5.33s
90:	learn: 0.3599302	test: 0.4267361	best: 0.4267361 (90)	total: 532ms	remaining: 5.31s
91:	learn: 0.3593131	test: 0.4263245	best: 0.4263245 (91)	total: 542ms	remaining: 5.35s
92:	learn: 0.3586683	test: 0.4256757	best: 0.4256757 (92)	total: 551ms	remaining: 5.37s
93:	learn: 0.3577304	test: 0.4255

174:	learn: 0.3159596	test: 0.4257876	best: 0.4228752 (126)	total: 945ms	remaining: 4.45s
175:	learn: 0.3157186	test: 0.4257429	best: 0.4228752 (126)	total: 950ms	remaining: 4.45s
176:	learn: 0.3154023	test: 0.4258077	best: 0.4228752 (126)	total: 961ms	remaining: 4.47s
177:	learn: 0.3152648	test: 0.4257873	best: 0.4228752 (126)	total: 966ms	remaining: 4.46s
178:	learn: 0.3142274	test: 0.4258862	best: 0.4228752 (126)	total: 975ms	remaining: 4.47s
179:	learn: 0.3134704	test: 0.4259951	best: 0.4228752 (126)	total: 980ms	remaining: 4.46s
180:	learn: 0.3134516	test: 0.4259894	best: 0.4228752 (126)	total: 982ms	remaining: 4.44s
181:	learn: 0.3132165	test: 0.4259895	best: 0.4228752 (126)	total: 986ms	remaining: 4.43s
182:	learn: 0.3130751	test: 0.4262678	best: 0.4228752 (126)	total: 991ms	remaining: 4.42s
183:	learn: 0.3127683	test: 0.4262772	best: 0.4228752 (126)	total: 996ms	remaining: 4.42s
184:	learn: 0.3125394	test: 0.4266352	best: 0.4228752 (126)	total: 1s	remaining: 4.41s
185:	learn: 0

279:	learn: 0.2774361	test: 0.4296332	best: 0.4228752 (126)	total: 1.51s	remaining: 3.88s
280:	learn: 0.2774126	test: 0.4296266	best: 0.4228752 (126)	total: 1.51s	remaining: 3.88s
281:	learn: 0.2767980	test: 0.4299712	best: 0.4228752 (126)	total: 1.52s	remaining: 3.87s
282:	learn: 0.2764561	test: 0.4298947	best: 0.4228752 (126)	total: 1.53s	remaining: 3.88s
283:	learn: 0.2756608	test: 0.4303256	best: 0.4228752 (126)	total: 1.54s	remaining: 3.87s
284:	learn: 0.2749564	test: 0.4307339	best: 0.4228752 (126)	total: 1.54s	remaining: 3.87s
285:	learn: 0.2744298	test: 0.4307972	best: 0.4228752 (126)	total: 1.55s	remaining: 3.86s
286:	learn: 0.2741147	test: 0.4311735	best: 0.4228752 (126)	total: 1.56s	remaining: 3.88s
287:	learn: 0.2738495	test: 0.4312950	best: 0.4228752 (126)	total: 1.57s	remaining: 3.88s
288:	learn: 0.2737606	test: 0.4313325	best: 0.4228752 (126)	total: 1.57s	remaining: 3.87s
289:	learn: 0.2737477	test: 0.4313400	best: 0.4228752 (126)	total: 1.58s	remaining: 3.88s
290:	learn

379:	learn: 0.2481738	test: 0.4334546	best: 0.4228752 (126)	total: 2.13s	remaining: 3.47s
380:	learn: 0.2476614	test: 0.4335830	best: 0.4228752 (126)	total: 2.13s	remaining: 3.46s
381:	learn: 0.2476349	test: 0.4335352	best: 0.4228752 (126)	total: 2.14s	remaining: 3.46s
382:	learn: 0.2474613	test: 0.4336261	best: 0.4228752 (126)	total: 2.14s	remaining: 3.45s
383:	learn: 0.2469138	test: 0.4337423	best: 0.4228752 (126)	total: 2.16s	remaining: 3.46s
384:	learn: 0.2467816	test: 0.4337678	best: 0.4228752 (126)	total: 2.17s	remaining: 3.46s
385:	learn: 0.2466530	test: 0.4337239	best: 0.4228752 (126)	total: 2.17s	remaining: 3.45s
386:	learn: 0.2463735	test: 0.4338983	best: 0.4228752 (126)	total: 2.18s	remaining: 3.45s
387:	learn: 0.2459441	test: 0.4340815	best: 0.4228752 (126)	total: 2.19s	remaining: 3.45s
388:	learn: 0.2458141	test: 0.4342740	best: 0.4228752 (126)	total: 2.19s	remaining: 3.44s
389:	learn: 0.2456948	test: 0.4341660	best: 0.4228752 (126)	total: 2.19s	remaining: 3.43s
390:	learn

481:	learn: 0.2196545	test: 0.4390275	best: 0.4228752 (126)	total: 2.73s	remaining: 2.93s
482:	learn: 0.2192856	test: 0.4388700	best: 0.4228752 (126)	total: 2.74s	remaining: 2.93s
483:	learn: 0.2187522	test: 0.4391336	best: 0.4228752 (126)	total: 2.74s	remaining: 2.92s
484:	learn: 0.2185448	test: 0.4391154	best: 0.4228752 (126)	total: 2.77s	remaining: 2.94s
485:	learn: 0.2183235	test: 0.4393638	best: 0.4228752 (126)	total: 2.77s	remaining: 2.93s
486:	learn: 0.2182866	test: 0.4393643	best: 0.4228752 (126)	total: 2.78s	remaining: 2.93s
487:	learn: 0.2181467	test: 0.4393147	best: 0.4228752 (126)	total: 2.79s	remaining: 2.92s
488:	learn: 0.2178584	test: 0.4394041	best: 0.4228752 (126)	total: 2.79s	remaining: 2.92s
489:	learn: 0.2175778	test: 0.4393243	best: 0.4228752 (126)	total: 2.8s	remaining: 2.92s
490:	learn: 0.2174541	test: 0.4393123	best: 0.4228752 (126)	total: 2.81s	remaining: 2.92s
491:	learn: 0.2170347	test: 0.4390908	best: 0.4228752 (126)	total: 2.82s	remaining: 2.92s
492:	learn:

575:	learn: 0.1960421	test: 0.4450762	best: 0.4228752 (126)	total: 3.55s	remaining: 2.61s
576:	learn: 0.1959064	test: 0.4448849	best: 0.4228752 (126)	total: 3.56s	remaining: 2.61s
577:	learn: 0.1956944	test: 0.4448974	best: 0.4228752 (126)	total: 3.58s	remaining: 2.61s
578:	learn: 0.1955444	test: 0.4449544	best: 0.4228752 (126)	total: 3.58s	remaining: 2.6s
579:	learn: 0.1954556	test: 0.4449154	best: 0.4228752 (126)	total: 3.59s	remaining: 2.6s
580:	learn: 0.1951996	test: 0.4450674	best: 0.4228752 (126)	total: 3.6s	remaining: 2.59s
581:	learn: 0.1951095	test: 0.4450782	best: 0.4228752 (126)	total: 3.6s	remaining: 2.59s
582:	learn: 0.1947638	test: 0.4451960	best: 0.4228752 (126)	total: 3.61s	remaining: 2.58s
583:	learn: 0.1947157	test: 0.4450622	best: 0.4228752 (126)	total: 3.63s	remaining: 2.59s
584:	learn: 0.1943385	test: 0.4449250	best: 0.4228752 (126)	total: 3.64s	remaining: 2.58s
585:	learn: 0.1941601	test: 0.4449053	best: 0.4228752 (126)	total: 3.64s	remaining: 2.57s
586:	learn: 0.

667:	learn: 0.1790864	test: 0.4508013	best: 0.4228752 (126)	total: 4.19s	remaining: 2.08s
668:	learn: 0.1788750	test: 0.4509723	best: 0.4228752 (126)	total: 4.2s	remaining: 2.08s
669:	learn: 0.1786641	test: 0.4512797	best: 0.4228752 (126)	total: 4.21s	remaining: 2.07s
670:	learn: 0.1785005	test: 0.4514805	best: 0.4228752 (126)	total: 4.21s	remaining: 2.07s
671:	learn: 0.1783931	test: 0.4518006	best: 0.4228752 (126)	total: 4.22s	remaining: 2.06s
672:	learn: 0.1781533	test: 0.4520323	best: 0.4228752 (126)	total: 4.22s	remaining: 2.05s
673:	learn: 0.1779299	test: 0.4522424	best: 0.4228752 (126)	total: 4.23s	remaining: 2.04s
674:	learn: 0.1779024	test: 0.4523473	best: 0.4228752 (126)	total: 4.23s	remaining: 2.04s
675:	learn: 0.1776945	test: 0.4527776	best: 0.4228752 (126)	total: 4.24s	remaining: 2.03s
676:	learn: 0.1772942	test: 0.4534139	best: 0.4228752 (126)	total: 4.25s	remaining: 2.03s
677:	learn: 0.1769885	test: 0.4533232	best: 0.4228752 (126)	total: 4.25s	remaining: 2.02s
678:	learn:

767:	learn: 0.1644094	test: 0.4580425	best: 0.4228752 (126)	total: 4.76s	remaining: 1.44s
768:	learn: 0.1642970	test: 0.4579873	best: 0.4228752 (126)	total: 4.77s	remaining: 1.43s
769:	learn: 0.1642126	test: 0.4580540	best: 0.4228752 (126)	total: 4.78s	remaining: 1.43s
770:	learn: 0.1639012	test: 0.4583269	best: 0.4228752 (126)	total: 4.79s	remaining: 1.42s
771:	learn: 0.1637704	test: 0.4584221	best: 0.4228752 (126)	total: 4.79s	remaining: 1.41s
772:	learn: 0.1637555	test: 0.4584566	best: 0.4228752 (126)	total: 4.8s	remaining: 1.41s
773:	learn: 0.1635864	test: 0.4587126	best: 0.4228752 (126)	total: 4.81s	remaining: 1.4s
774:	learn: 0.1635541	test: 0.4587392	best: 0.4228752 (126)	total: 4.82s	remaining: 1.4s
775:	learn: 0.1634548	test: 0.4589000	best: 0.4228752 (126)	total: 4.82s	remaining: 1.39s
776:	learn: 0.1633090	test: 0.4587799	best: 0.4228752 (126)	total: 4.82s	remaining: 1.38s
777:	learn: 0.1632821	test: 0.4588112	best: 0.4228752 (126)	total: 4.83s	remaining: 1.38s
778:	learn: 0

872:	learn: 0.1516000	test: 0.4627294	best: 0.4228752 (126)	total: 5.35s	remaining: 779ms
873:	learn: 0.1514207	test: 0.4631149	best: 0.4228752 (126)	total: 5.37s	remaining: 775ms
874:	learn: 0.1513887	test: 0.4631541	best: 0.4228752 (126)	total: 5.38s	remaining: 769ms
875:	learn: 0.1513158	test: 0.4632672	best: 0.4228752 (126)	total: 5.39s	remaining: 763ms
876:	learn: 0.1513102	test: 0.4632869	best: 0.4228752 (126)	total: 5.39s	remaining: 756ms
877:	learn: 0.1509040	test: 0.4637547	best: 0.4228752 (126)	total: 5.4s	remaining: 750ms
878:	learn: 0.1507414	test: 0.4643722	best: 0.4228752 (126)	total: 5.4s	remaining: 744ms
879:	learn: 0.1504710	test: 0.4648010	best: 0.4228752 (126)	total: 5.41s	remaining: 738ms
880:	learn: 0.1503907	test: 0.4651051	best: 0.4228752 (126)	total: 5.41s	remaining: 731ms
881:	learn: 0.1502966	test: 0.4650427	best: 0.4228752 (126)	total: 5.42s	remaining: 725ms
882:	learn: 0.1502895	test: 0.4650460	best: 0.4228752 (126)	total: 5.43s	remaining: 719ms
883:	learn: 

977:	learn: 0.1384744	test: 0.4714006	best: 0.4228752 (126)	total: 5.96s	remaining: 134ms
978:	learn: 0.1383864	test: 0.4713636	best: 0.4228752 (126)	total: 5.96s	remaining: 128ms
979:	learn: 0.1382769	test: 0.4715964	best: 0.4228752 (126)	total: 5.97s	remaining: 122ms
980:	learn: 0.1380082	test: 0.4718158	best: 0.4228752 (126)	total: 5.98s	remaining: 116ms
981:	learn: 0.1379374	test: 0.4717441	best: 0.4228752 (126)	total: 5.98s	remaining: 110ms
982:	learn: 0.1377881	test: 0.4718173	best: 0.4228752 (126)	total: 5.99s	remaining: 104ms
983:	learn: 0.1376890	test: 0.4717485	best: 0.4228752 (126)	total: 5.99s	remaining: 97.4ms
984:	learn: 0.1376150	test: 0.4719544	best: 0.4228752 (126)	total: 6s	remaining: 91.4ms
985:	learn: 0.1375469	test: 0.4720366	best: 0.4228752 (126)	total: 6.01s	remaining: 85.3ms
986:	learn: 0.1374755	test: 0.4723414	best: 0.4228752 (126)	total: 6.01s	remaining: 79.2ms
987:	learn: 0.1373501	test: 0.4723450	best: 0.4228752 (126)	total: 6.02s	remaining: 73.1ms
988:	lea

89:	learn: 0.3635493	test: 0.4467627	best: 0.4467627 (89)	total: 441ms	remaining: 4.46s
90:	learn: 0.3625946	test: 0.4467624	best: 0.4467624 (90)	total: 457ms	remaining: 4.56s
91:	learn: 0.3619440	test: 0.4462825	best: 0.4462825 (91)	total: 462ms	remaining: 4.56s
92:	learn: 0.3609993	test: 0.4463020	best: 0.4462825 (91)	total: 466ms	remaining: 4.55s
93:	learn: 0.3601188	test: 0.4460179	best: 0.4460179 (93)	total: 471ms	remaining: 4.54s
94:	learn: 0.3598265	test: 0.4460474	best: 0.4460179 (93)	total: 476ms	remaining: 4.53s
95:	learn: 0.3589359	test: 0.4457914	best: 0.4457914 (95)	total: 480ms	remaining: 4.52s
96:	learn: 0.3581293	test: 0.4448283	best: 0.4448283 (96)	total: 486ms	remaining: 4.52s
97:	learn: 0.3579063	test: 0.4449543	best: 0.4448283 (96)	total: 489ms	remaining: 4.5s
98:	learn: 0.3568947	test: 0.4443846	best: 0.4443846 (98)	total: 495ms	remaining: 4.5s
99:	learn: 0.3560971	test: 0.4442542	best: 0.4442542 (99)	total: 502ms	remaining: 4.52s
100:	learn: 0.3560473	test: 0.4441

186:	learn: 0.3185135	test: 0.4336919	best: 0.4335806 (182)	total: 902ms	remaining: 3.92s
187:	learn: 0.3182879	test: 0.4334746	best: 0.4334746 (187)	total: 908ms	remaining: 3.92s
188:	learn: 0.3176241	test: 0.4335242	best: 0.4334746 (187)	total: 913ms	remaining: 3.92s
189:	learn: 0.3174486	test: 0.4337107	best: 0.4334746 (187)	total: 919ms	remaining: 3.92s
190:	learn: 0.3166886	test: 0.4338364	best: 0.4334746 (187)	total: 923ms	remaining: 3.91s
191:	learn: 0.3164893	test: 0.4339585	best: 0.4334746 (187)	total: 928ms	remaining: 3.9s
192:	learn: 0.3158040	test: 0.4339656	best: 0.4334746 (187)	total: 933ms	remaining: 3.9s
193:	learn: 0.3153715	test: 0.4338474	best: 0.4334746 (187)	total: 938ms	remaining: 3.9s
194:	learn: 0.3147256	test: 0.4337780	best: 0.4334746 (187)	total: 943ms	remaining: 3.89s
195:	learn: 0.3138405	test: 0.4331794	best: 0.4331794 (195)	total: 947ms	remaining: 3.88s
196:	learn: 0.3138212	test: 0.4331767	best: 0.4331767 (196)	total: 949ms	remaining: 3.87s
197:	learn: 0

286:	learn: 0.2812224	test: 0.4353356	best: 0.4329959 (200)	total: 1.45s	remaining: 3.61s
287:	learn: 0.2805266	test: 0.4349236	best: 0.4329959 (200)	total: 1.46s	remaining: 3.62s
288:	learn: 0.2800692	test: 0.4346568	best: 0.4329959 (200)	total: 1.48s	remaining: 3.63s
289:	learn: 0.2799482	test: 0.4346233	best: 0.4329959 (200)	total: 1.48s	remaining: 3.62s
290:	learn: 0.2797624	test: 0.4346360	best: 0.4329959 (200)	total: 1.49s	remaining: 3.62s
291:	learn: 0.2797033	test: 0.4346608	best: 0.4329959 (200)	total: 1.49s	remaining: 3.62s
292:	learn: 0.2792967	test: 0.4346785	best: 0.4329959 (200)	total: 1.51s	remaining: 3.64s
293:	learn: 0.2787014	test: 0.4348633	best: 0.4329959 (200)	total: 1.52s	remaining: 3.66s
294:	learn: 0.2783607	test: 0.4346658	best: 0.4329959 (200)	total: 1.53s	remaining: 3.67s
295:	learn: 0.2778161	test: 0.4345042	best: 0.4329959 (200)	total: 1.54s	remaining: 3.67s
296:	learn: 0.2775889	test: 0.4344698	best: 0.4329959 (200)	total: 1.55s	remaining: 3.67s
297:	learn

384:	learn: 0.2520435	test: 0.4375513	best: 0.4329959 (200)	total: 2.48s	remaining: 3.96s
385:	learn: 0.2517899	test: 0.4378531	best: 0.4329959 (200)	total: 2.49s	remaining: 3.96s
386:	learn: 0.2515006	test: 0.4377718	best: 0.4329959 (200)	total: 2.5s	remaining: 3.96s
387:	learn: 0.2512379	test: 0.4377119	best: 0.4329959 (200)	total: 2.5s	remaining: 3.95s
388:	learn: 0.2510581	test: 0.4377262	best: 0.4329959 (200)	total: 2.51s	remaining: 3.95s
389:	learn: 0.2509362	test: 0.4375457	best: 0.4329959 (200)	total: 2.52s	remaining: 3.95s
390:	learn: 0.2504774	test: 0.4383473	best: 0.4329959 (200)	total: 2.54s	remaining: 3.95s
391:	learn: 0.2501094	test: 0.4382562	best: 0.4329959 (200)	total: 2.54s	remaining: 3.95s
392:	learn: 0.2496373	test: 0.4383610	best: 0.4329959 (200)	total: 2.56s	remaining: 3.96s
393:	learn: 0.2494219	test: 0.4384710	best: 0.4329959 (200)	total: 2.57s	remaining: 3.96s
394:	learn: 0.2493828	test: 0.4384946	best: 0.4329959 (200)	total: 2.58s	remaining: 3.96s
395:	learn: 

483:	learn: 0.2280149	test: 0.4446771	best: 0.4329959 (200)	total: 3.17s	remaining: 3.38s
484:	learn: 0.2279065	test: 0.4446978	best: 0.4329959 (200)	total: 3.18s	remaining: 3.38s
485:	learn: 0.2278488	test: 0.4447079	best: 0.4329959 (200)	total: 3.19s	remaining: 3.37s
486:	learn: 0.2275953	test: 0.4447642	best: 0.4329959 (200)	total: 3.2s	remaining: 3.37s
487:	learn: 0.2274492	test: 0.4448738	best: 0.4329959 (200)	total: 3.21s	remaining: 3.36s
488:	learn: 0.2272707	test: 0.4450967	best: 0.4329959 (200)	total: 3.21s	remaining: 3.35s
489:	learn: 0.2270391	test: 0.4452228	best: 0.4329959 (200)	total: 3.21s	remaining: 3.35s
490:	learn: 0.2270044	test: 0.4452338	best: 0.4329959 (200)	total: 3.22s	remaining: 3.34s
491:	learn: 0.2269601	test: 0.4451865	best: 0.4329959 (200)	total: 3.23s	remaining: 3.33s
492:	learn: 0.2266004	test: 0.4450422	best: 0.4329959 (200)	total: 3.23s	remaining: 3.32s
493:	learn: 0.2265431	test: 0.4450627	best: 0.4329959 (200)	total: 3.23s	remaining: 3.31s
494:	learn:

575:	learn: 0.2062496	test: 0.4532716	best: 0.4329959 (200)	total: 3.79s	remaining: 2.79s
576:	learn: 0.2062151	test: 0.4533438	best: 0.4329959 (200)	total: 3.8s	remaining: 2.79s
577:	learn: 0.2058088	test: 0.4534146	best: 0.4329959 (200)	total: 3.81s	remaining: 2.78s
578:	learn: 0.2055292	test: 0.4534992	best: 0.4329959 (200)	total: 3.81s	remaining: 2.77s
579:	learn: 0.2053731	test: 0.4534799	best: 0.4329959 (200)	total: 3.82s	remaining: 2.77s
580:	learn: 0.2053213	test: 0.4534697	best: 0.4329959 (200)	total: 3.82s	remaining: 2.76s
581:	learn: 0.2051173	test: 0.4536006	best: 0.4329959 (200)	total: 3.83s	remaining: 2.75s
582:	learn: 0.2048930	test: 0.4537239	best: 0.4329959 (200)	total: 3.84s	remaining: 2.74s
583:	learn: 0.2047132	test: 0.4536814	best: 0.4329959 (200)	total: 3.84s	remaining: 2.74s
584:	learn: 0.2043622	test: 0.4538858	best: 0.4329959 (200)	total: 3.85s	remaining: 2.73s
585:	learn: 0.2034866	test: 0.4535530	best: 0.4329959 (200)	total: 3.86s	remaining: 2.73s
586:	learn:

672:	learn: 0.1851233	test: 0.4570304	best: 0.4329959 (200)	total: 4.4s	remaining: 2.14s
673:	learn: 0.1847874	test: 0.4572075	best: 0.4329959 (200)	total: 4.41s	remaining: 2.13s
674:	learn: 0.1845932	test: 0.4576062	best: 0.4329959 (200)	total: 4.43s	remaining: 2.13s
675:	learn: 0.1845698	test: 0.4576281	best: 0.4329959 (200)	total: 4.44s	remaining: 2.13s
676:	learn: 0.1841689	test: 0.4574993	best: 0.4329959 (200)	total: 4.45s	remaining: 2.12s
677:	learn: 0.1837736	test: 0.4582684	best: 0.4329959 (200)	total: 4.46s	remaining: 2.12s
678:	learn: 0.1834100	test: 0.4582435	best: 0.4329959 (200)	total: 4.47s	remaining: 2.11s
679:	learn: 0.1833353	test: 0.4583824	best: 0.4329959 (200)	total: 4.47s	remaining: 2.1s
680:	learn: 0.1831506	test: 0.4589973	best: 0.4329959 (200)	total: 4.48s	remaining: 2.1s
681:	learn: 0.1828841	test: 0.4596078	best: 0.4329959 (200)	total: 4.48s	remaining: 2.09s
682:	learn: 0.1823638	test: 0.4596259	best: 0.4329959 (200)	total: 4.49s	remaining: 2.08s
683:	learn: 0

773:	learn: 0.1651828	test: 0.4662906	best: 0.4329959 (200)	total: 5.33s	remaining: 1.56s
774:	learn: 0.1649447	test: 0.4670887	best: 0.4329959 (200)	total: 5.34s	remaining: 1.55s
775:	learn: 0.1649126	test: 0.4671313	best: 0.4329959 (200)	total: 5.35s	remaining: 1.54s
776:	learn: 0.1647330	test: 0.4671245	best: 0.4329959 (200)	total: 5.36s	remaining: 1.54s
777:	learn: 0.1641826	test: 0.4665487	best: 0.4329959 (200)	total: 5.36s	remaining: 1.53s
778:	learn: 0.1640221	test: 0.4666968	best: 0.4329959 (200)	total: 5.37s	remaining: 1.52s
779:	learn: 0.1639610	test: 0.4666884	best: 0.4329959 (200)	total: 5.38s	remaining: 1.52s
780:	learn: 0.1639345	test: 0.4666411	best: 0.4329959 (200)	total: 5.38s	remaining: 1.51s
781:	learn: 0.1637145	test: 0.4664733	best: 0.4329959 (200)	total: 5.39s	remaining: 1.5s
782:	learn: 0.1636122	test: 0.4665241	best: 0.4329959 (200)	total: 5.39s	remaining: 1.5s
783:	learn: 0.1635082	test: 0.4666500	best: 0.4329959 (200)	total: 5.4s	remaining: 1.49s
784:	learn: 0

883:	learn: 0.1484172	test: 0.4796165	best: 0.4329959 (200)	total: 6.15s	remaining: 808ms
884:	learn: 0.1483751	test: 0.4796209	best: 0.4329959 (200)	total: 6.16s	remaining: 801ms
885:	learn: 0.1481575	test: 0.4789745	best: 0.4329959 (200)	total: 6.18s	remaining: 795ms
886:	learn: 0.1481397	test: 0.4790440	best: 0.4329959 (200)	total: 6.19s	remaining: 789ms
887:	learn: 0.1479454	test: 0.4792794	best: 0.4329959 (200)	total: 6.2s	remaining: 781ms
888:	learn: 0.1477090	test: 0.4797078	best: 0.4329959 (200)	total: 6.2s	remaining: 774ms
889:	learn: 0.1475559	test: 0.4796615	best: 0.4329959 (200)	total: 6.21s	remaining: 767ms
890:	learn: 0.1473669	test: 0.4797403	best: 0.4329959 (200)	total: 6.21s	remaining: 760ms
891:	learn: 0.1472550	test: 0.4795886	best: 0.4329959 (200)	total: 6.22s	remaining: 753ms
892:	learn: 0.1470853	test: 0.4794652	best: 0.4329959 (200)	total: 6.22s	remaining: 746ms
893:	learn: 0.1469820	test: 0.4794814	best: 0.4329959 (200)	total: 6.23s	remaining: 739ms
894:	learn: 

993:	learn: 0.1348787	test: 0.4858125	best: 0.4329959 (200)	total: 6.98s	remaining: 42.1ms
994:	learn: 0.1346008	test: 0.4859523	best: 0.4329959 (200)	total: 6.99s	remaining: 35.1ms
995:	learn: 0.1343902	test: 0.4860028	best: 0.4329959 (200)	total: 6.99s	remaining: 28.1ms
996:	learn: 0.1342712	test: 0.4859343	best: 0.4329959 (200)	total: 7s	remaining: 21.1ms
997:	learn: 0.1342389	test: 0.4859595	best: 0.4329959 (200)	total: 7.01s	remaining: 14ms
998:	learn: 0.1341691	test: 0.4859469	best: 0.4329959 (200)	total: 7.01s	remaining: 7.02ms
999:	learn: 0.1340561	test: 0.4858727	best: 0.4329959 (200)	total: 7.02s	remaining: 0us

bestTest = 0.4329958933
bestIteration = 200



In [10]:
print('Best validation accuracy score: {:.2f}±{:.2f} on step {}'.format(
    np.max(cv_data['test-Accuracy-mean']),
    cv_data['test-Accuracy-std'][np.argmax(cv_data['test-Accuracy-mean'])],
    np.argmax(cv_data['test-Accuracy-mean'])
))

Best validation accuracy score: 0.82±0.01 on step 897


In [11]:
print('Simple model validation accuracy: {:.4}'.format(
    accuracy_score(y_validation, model.predict(X_validation))
))

Simple model validation accuracy: 0.8161


## 3 Using the Model

Let's apply the model to data in order to predict the labels. Also, we will have a look at CatBoost's feature importance.

In [12]:
# This will give the actual predictions
predictions = model.predict(X_test)
print(predictions[:10])


[0 0 0 0 1 0 1 0 1 0]


In [13]:
# And this the underlying probability values
predictions_probs = model.predict_proba(X_test)
print(predictions_probs[:10])

[[0.85473931 0.14526069]
 [0.76313031 0.23686969]
 [0.88972889 0.11027111]
 [0.87876173 0.12123827]
 [0.3611047  0.6388953 ]
 [0.90513381 0.09486619]
 [0.33434185 0.66565815]
 [0.78468564 0.21531436]
 [0.39429048 0.60570952]
 [0.94047549 0.05952451]]


### 3.1 Best Model
We can have a look at the difference between the simple CatBoost model and choosing the best model during the training iteration.

In [14]:
params = {
    'iterations': 500,
    'learning_rate': 0.1,
    'eval_metric': metrics.Accuracy(),
    'random_seed': 42,
    'logging_level': 'Silent',
    'use_best_model': False
}
train_pool = Pool(X_train, y_train, cat_features=categorical_features_indices)
validate_pool = Pool(X_validation, y_validation, cat_features=categorical_features_indices)

In [15]:
model = CatBoostClassifier(**params)
model.fit(train_pool, eval_set=validate_pool)

best_model_params = params.copy()
best_model_params.update({
    'use_best_model': True
})
best_model = CatBoostClassifier(**best_model_params)
best_model.fit(train_pool, eval_set=validate_pool);

print('Simple model validation accuracy: {:.4}'.format(
    accuracy_score(y_validation, model.predict(X_validation))
))
print('')

print('Best model validation accuracy: {:.4}'.format(
    accuracy_score(y_validation, best_model.predict(X_validation))
))


Simple model validation accuracy: 0.7982

Best model validation accuracy: 0.8251


So there is a noticeable increase in accuracy on the validation set.

### Feature Importance

We will use CatBoost's feature ranking to see how the models differ. In a more complicated task with many more features, this could be very useful to interpret the model.

In [16]:
feature_importances = model.get_feature_importance(train_pool)
feature_names = X_train.columns
print("Feature importances for the baseline model")
for score, name in sorted(zip(feature_importances, feature_names), reverse=True):
    print('{}: {}'.format(name, score))

Feature importances for the baseline model
Sex: 21.59911877701903
Age: 19.398416335821832
Fare: 13.886969720468315
Pclass: 13.770872263093882
Embarked: 7.030284738986945
Ticket: 6.964771018892179
Parch: 6.355643060218354
SibSp: 6.010444835646993
Cabin: 4.983479249852257
PassengerId: 0.0
Name: 0.0


In [17]:
feature_importances = best_model.get_feature_importance(train_pool)
feature_names = X_train.columns
print("Feature importances for the best model")
for score, name in sorted(zip(feature_importances, feature_names), reverse=True):
    print('{}: {}'.format(name, score))

Feature importances for the best model
Sex: 42.17474818544473
Pclass: 16.294343982697637
Ticket: 9.797084805047552
Parch: 7.53610285333354
Age: 7.041597602831637
Fare: 4.9866064486621715
Cabin: 4.836829329849025
Embarked: 3.6878977676281557
SibSp: 3.6447890245054957
PassengerId: 0.0
Name: 0.0


PassengerId and Name have absolutely no effect on the outcome - which is good because there is no reason why they should. Compared to the aggregated baseline model, the best model attributes much more importance to the feature "Sex", while the importances of the features "Age" and "Fare" have been reduced noticeably.